pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_controller.go

```diff
@@ -0,0 +1,231 @@
+
+		queue: queue,
+
+		createAPIDefinition: createAPIDefinition,
+
+		apiSets: map[dynamiccontext.APIDomainKey]apidefinition.APIDefinitionSet{},
+	}
+
+	if err := apiExportInformer.Informer().AddIndexers(cache.Indexers{
+		byWorkspace: indexByWorksapce,
+	}); err != nil {
+		return nil, err
+	}
+
+	apiExportInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+	})
+
+	apiResourceSchemaInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIResourceSchema(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIResourceSchema(obj)
+		},
+	})
+
+	return c, nil
+}
+
+// APIReconciler is a controller that reconciles APIResourceImport resources
+// and delegates the corresponding WorkloadClusterAPI management to the given WorkloadClusterAPIManager.
+type APIReconciler struct {
+	kcpClusterClient kcpclient.ClusterInterface
+
+	apiResourceSchemaLister  apislisters.APIResourceSchemaLister
+	apiResourceSchemaIndexer cache.Indexer
+
+	apiExportLister  apislisters.APIExportLister
+	apiExportIndexer cache.Indexer
+
+	queue workqueue.RateLimitingInterface
+
+	createAPIDefinition CreateAPIDefinitionFunc
+
+	mutex   sync.RWMutex // protects the map, not the values!
+	apiSets map[dynamiccontext.APIDomainKey]apidefinition.APIDefinitionSet
+}
+
+func (c *APIReconciler) enqueueAPIResourceSchema(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	clusterName, name := clusters.SplitClusterAwareKey(key)
+	exports, err := c.apiExportIndexer.ByIndex(byWorkspace, clusterName.String())
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	if len(exports) == 0 {
+		klog.V(2).Infof("No kubernetes APIExport found for APIResourceSchema %s|%s", clusterName, name)
+		return
+	}
+
+	for _, obj := range exports {
+		export := obj.(*apisv1alpha1.APIExport)
+		klog.V(2).Infof("Queueing APIExport %s|%s for APIResourceSchema %s", clusterName, export.Name, name)
+		c.enqueueAPIExport(obj)
+	}
+}
+
+func (c *APIReconciler) enqueueAPIExport(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *APIReconciler) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *APIReconciler) Start(ctx context.Context) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
+	defer klog.Infof("Shutting down %s controller", ControllerName)
+
+	go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+
+	// stop all watches if the controller is stopped
```

Comment:
Does this somehow need to get plumbed into the storage's Destroy?

Reply:
we pass through the channel already (you reminded me this week). So probably not?!